ec/ecresource/deploymentresource/elasticsearch/v2/node_roles.go

```diff
@@ -0,0 +1,203 @@
+// migrated over to node_roles. Which will be true when:
+// * The version field doesn't change.
+// * The version field changes but:
+//   - The Elasticsearch.0.toplogy doesn't have any node_type_* set.
+func legacyToNodeRoles(ctx context.Context, stateVersion, planVersion types.String, planElasticsearch types.Object) (bool, diag.Diagnostics) {
+	if stateVersion.Value == "" || stateVersion.Value == planVersion.Value {
+		return true, nil
+	}
+
+	var diags diag.Diagnostics
+	oldVersion, err := semver.Parse(stateVersion.Value)
+	if err != nil {
+		diags.AddError("Failed to parse previous Elasticsearch version", err.Error())
+		return false, diags
+	}
+	newVersion, err := semver.Parse(planVersion.Value)
+	if err != nil {
+		diags.AddError("Failed to parse new Elasticsearch version", err.Error())
+		return false, diags
+	}
+
+	// if the version change moves from non-node_roles to one
+	// that supports node roles, do not migrate on that step.
+	if oldVersion.LT(utils.DataTiersVersion) && newVersion.GE(utils.DataTiersVersion) {
+		return false, nil
+	}
+
+	// When any topology elements in the state have the node_type_*
+	// properties set, the node_role field cannot be used, since
+	// we'd be changing the version AND migrating over `node_role`s
+	// which is not permitted by the API.
+
+	var es *ElasticsearchTF
+
+	if diags := tfsdk.ValueAs(ctx, planElasticsearch, &es); diags.HasError() {
+		return false, diags
+	}
+
+	if es == nil {
+		diags.AddError("Cannot migrate node types to node roles", "cannot find elasticsearch object")
+		return false, diags
+	}
+
+	tiers, diags := es.topologies(ctx)
+
+	if diags.HasError() {
+		return false, diags
+	}
+
+	for _, tier := range tiers {
+		if tier != nil && tier.HasNodeType() {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func useStateAndNodeRolesInPlanModifiers(ctx context.Context, req tfsdk.ModifyAttributePlanRequest, resp *tfsdk.ModifyAttributePlanResponse) (useState, useNodeRoles bool) {
+	if req.AttributeState == nil || resp.AttributePlan == nil || req.AttributeConfig == nil {
+		return false, false
+	}
+
+	if !resp.AttributePlan.IsUnknown() {
+		return false, false
+	}
+
+	// if the config is the unknown value, use the unknown value otherwise, interpolation gets messed up
+	// it's the precaution taken from the Framework's `UseStateForUnknown` plan modifier
+	if req.AttributeConfig.IsUnknown() {
+		return false, false
+	}
+
+	// if there is no state for "version" return
+	var stateVersion types.String
+
+	if diags := req.State.GetAttribute(ctx, path.Root("version"), &stateVersion); diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return false, false
+	}
+
+	if stateVersion.IsNull() {
+		return false, false
+	}
+
+	// if template changed return
+	templateChanged, diags := attributeChanged(ctx, path.Root("deployment_template_id"), req)
+
+	resp.Diagnostics.Append(diags...)
+
+	if diags.HasError() {
+		return false, false
+	}
+
+	if templateChanged {
+		return false, false
+	}
```

Comment:
The function is used only in `node_roles` and `node_types` plan modifies. The `if` case prevents the plan modifies from using the current state as a planned value. Without that, once the template is modified, the plan modifier would still use the current state. I think it's more about precaution - once the template is changed, the plan output suggests that future values for `node_roles` and `node_types` are "known after apply". They still can be the same as the current state values though. The problem with plan modifies is that if they make wrong decision, it becomes an error in the provider while if they doesn't suggest any value (effectively keeping "known after apply"), it could be just extra output in the plan.